Evaluation_1/TPU_Samples/image_retraining_tpu.py

```diff
@@ -0,0 +1,125 @@
+    global FEATURE_SHAPE
+    assert FEATURE_SHAPE is not None
+    feature = tf.compat.v1.placeholder(tf.float32, shape=[None] + FEATURE_SHAPE, name="input")
+    reciever_tensor = {"input": feature}
+    return tf.estimator.export.ServingInputReciver(feature, reciever_tensor)
+def model_fn(features, labels, mode, params):
+    global NUM_CLASSES
+    assert NUM_CLASSES is not None
+    model = tf.keras.Sequential([
+      hub.KerasLayer("https://tfhub.dev/google/tf2-preview/inception_v3/feature_vector/4",
+        output_shape=[2048],
+        trainable=False
+      ),
+      tf.keras.layers.Dense(NUM_CLASSES, activation="softmax")
+    ])
+    optimizer = None
+    if mode == tf.estimator.ModeKeys.TRAIN:
+        optimizer = tf.optimizers.Adam(params.get("learing_rate", 1e-3))
+        if params.get("use_tpu", True):
+          optimizer = tpu_optimizer.CrossShardOptimizer(optimizer)
+
+    with tf.GradientTape() as tape:
+        logits = model(features)
+        if mode == tf.estimator.ModeKeys.PREDICT:
+            preds = {
+                "predictions": logits
+            }
+            return tpu_estimator.TPUEstimatorSpec(mode, predictions=preds)
+        loss = tf.keras.losses.SparseCategoricalCrossentropy(
+            from_logits=True)(labels, logits)
+    if mode == tf.estimator.ModeKeys.EVAL:
+        return tpu_estimator.TPUEstimatorSpec(mode, loss=loss)
+
+    def train_fn():
+        assert optimizer is not None
+        gradient = tape.gradient(loss, model.trainable_variables)
+        global_step = tf.compat.v1.train.get_global_step()
+        update_global_step = tf.compat.v1.assign(global_step, global_step + 1, name='update_global_step')
+        with tf.control_dependencies([update_global_step]):
+          apply_grads = optimizer.apply_gradients(zip(gradient, model.trainable_variables))
+        return apply_grads
+
+    if mode == tf.estimator.ModeKeys.TRAIN:
+        return tpu_estimator.TPUEstimatorSpec(mode, loss=loss, train_op=train_fn())
+
+def main(_):
+  os.environ["TFHUB_CACHE_DIR"] = os.path.join(FLAGS.model_dir,"tfhub_modules")
+  os.environ["TFHUB_DOWNLOAD_PROGRESS"] = "True"
+  input_fn = partial(input_, iterations=FLAGS.iterations)
+  cluster = tf.distribute.cluster_resolver.TPUClusterResolver(tpu=FLAGS.tpu)
+  run_config =  tpu_config.RunConfig(
+    model_dir = FLAGS.model_dir,
+    cluster=cluster,
+    tpu_config=tpu_config.TPUConfig(FLAGS.iterations))
+
+  classifier = tpu_estimator.TPUEstimator(
+    model_fn=model_fn,
+    use_tpu=FLAGS.use_tpu,
+    train_batch_size=FLAGS.batch_size,
+    eval_batch_size=FLAGS.batch_size,
+    config=run_config,
+    params={
+      "use_tpu": FLAGS.use_tpu,
+      "data_dir": FLAGS.data_dir,
+      "dataset": FLAGS.dataset
+    }
+  )
+
+  classifier.train(
+          input_fn=lambda params:input_fn(
+              mode=tf.estimator.ModeKeys.TRAIN,
+              **params),
+          max_steps=1000)#.evaluate(
+                  #input_fn=lambda params:input_fn(
```

Comment:
Remove commented-out code?
If its something that will be enabled later, add a TODO (see other comment)

Reply:
Done